From: Bjoern Doebel <doebel@amazon.de>
Date: Mon, 25 Nov 2024 04:46:13 +0000
Subject: LU-17887 obd: do not update obd_memory from RCU

OBD_FREE_PRE() should not be run from an RCU
callback as the obdclass module may have been
unloaded during the RCU grace period.

Signed-off-by: Bruno Faccini <bfaccini@nvidia.com>
Change-Id: I6f663b2aed2e60c15f2a1b9755b2c4050bd91ce2
Reviewed-on: https://review.whamcloud.com/c/fs/lustre-release/+/55263
Reviewed-by: Alex Zhuravlev <bzzz@whamcloud.com>
Reviewed-by: James Simmons <jsimmons@infradead.org>
Reviewed-by: Oleg Drokin <green@whamcloud.com>
Tested-by: jenkins <devops@whamcloud.com>
Tested-by: Maloo <maloo@whamcloud.com>

Signed-off-by: Bjoern Doebel <doebel@amazon.de>
---
 .../staging/lustrefsx/lustre/obdclass/lu_object.c  | 14 ++++----------
 1 file changed, 4 insertions(+), 10 deletions(-)

diff --git a/drivers/staging/lustrefsx/lustre/obdclass/lu_object.c b/drivers/staging/lustrefsx/lustre/obdclass/lu_object.c
--- a/drivers/staging/lustrefsx/lustre/obdclass/lu_object.c
+++ b/drivers/staging/lustrefsx/lustre/obdclass/lu_object.c
@@ -2034,14 +2034,6 @@ int lu_env_add(struct lu_env *env)
 }
 EXPORT_SYMBOL(lu_env_add);
 
-static void lu_env_item_free(struct rcu_head *head)
-{
-	struct lu_env_item *lei;
-
-	lei = container_of(head, struct lu_env_item, lei_rcu_head);
-	OBD_FREE_PTR(lei);
-}
-
 void lu_env_remove(struct lu_env *env)
 {
 	struct lu_env_item *lei;
@@ -2064,8 +2056,10 @@ void lu_env_remove(struct lu_env *env)
 	lei = rhashtable_lookup_fast(&lu_env_rhash, &task,
 				     lu_env_rhash_params);
 	if (lei && rhashtable_remove_fast(&lu_env_rhash, &lei->lei_linkage,
-					  lu_env_rhash_params) == 0)
-		call_rcu(&lei->lei_rcu_head, lu_env_item_free);
+									  lu_env_rhash_params) == 0) {
+		OBD_FREE_PRE(lei, sizeof(*lei), "kfreed");
+		kfree_rcu(lei, lei_rcu_head);
+	}
 }
 EXPORT_SYMBOL(lu_env_remove);
 

